Mathlib/LinearAlgebra/Eigenspace/Basic.lean

```diff
@@ -379,13 +268,136 @@ theorem generalizedEigenspace_le_generalizedEigenspace_finrank [FiniteDimensiona
+  simp only [SetLike.mem_coe, mem_generalizedEigenspace] at hx ⊢
+  rw [← LinearMap.comp_apply, ← LinearMap.mul_eq_comp, h.eq, LinearMap.mul_eq_comp,
+    LinearMap.comp_apply, hx, map_zero]
+
+lemma mapsTo_iSup_generalizedEigenspace_of_comm {f g : End R M} (h : Commute f g) (μ : R) :
+    MapsTo g ↑(⨆ k, f.generalizedEigenspace μ k) ↑(⨆ k, f.generalizedEigenspace μ k) := by
+  intro x hx
+  simp only [Submodule.coe_iSup_of_chain, mem_iUnion, SetLike.mem_coe] at hx
+  obtain ⟨k, hk : x ∈ f.generalizedEigenspace μ k⟩ := hx
+  simp only [Submodule.coe_iSup_of_chain, mem_iUnion, SetLike.mem_coe]
+  exact ⟨k, f.mapsTo_generalizedEigenspace_of_comm h μ k hk⟩
+
+lemma disjoint_generalizedEigenspace [NoZeroSMulDivisors R M]
+    (f : End R M) {μ₁ μ₂ : R} (hμ : μ₁ ≠ μ₂) (k l : ℕ) :
+    Disjoint (f.generalizedEigenspace μ₁ k) (f.generalizedEigenspace μ₂ l) := by
+  nontriviality M
+  have := NoZeroSMulDivisors.isReduced R M
+  rw [disjoint_iff]
+  set p := f.generalizedEigenspace μ₁ k ⊓ f.generalizedEigenspace μ₂ l
+  by_contra hp
+  replace hp : Nontrivial p := Submodule.nontrivial_iff_ne_bot.mpr hp
+  have hc₁ : Commute f (f - algebraMap R (End R M) μ₁) :=
+    Commute.sub_right rfl (Algebra.commute_algebraMap_right μ₁ f)
+  have hc₂ : Commute f (f - algebraMap R (End R M) μ₂) :=
+    Commute.sub_right rfl (Algebra.commute_algebraMap_right μ₂ f)
+  let f₁ : End R p := LinearMap.restrict (f - algebraMap R (End R M) μ₁) (MapsTo.inter_inter
+    (mapsTo_generalizedEigenspace_of_comm hc₁ μ₁ k) (mapsTo_generalizedEigenspace_of_comm hc₁ μ₂ l))
+  let f₂ : End R p := LinearMap.restrict (f - algebraMap R (End R M) μ₂) (MapsTo.inter_inter
+    (mapsTo_generalizedEigenspace_of_comm hc₂ μ₁ k) (mapsTo_generalizedEigenspace_of_comm hc₂ μ₂ l))
+  have hn₁ : IsNilpotent f₁ := by
+    use k; ext ⟨x, hx⟩; simpa [LinearMap.restrict_apply, LinearMap.pow_restrict _] using hx.1
+  have hn₂ : IsNilpotent f₂ := by
+    use l; ext ⟨x, hx⟩; simpa [LinearMap.restrict_apply, LinearMap.pow_restrict _] using hx.2
+  have : IsNilpotent (f₂ - f₁) := by
+    apply Commute.isNilpotent_sub _ hn₂ hn₁
+    ext; simp [smul_sub, sub_sub, smul_comm μ₁, add_sub_left_comm]
+  have hf₁₂ : f₂ - f₁ = algebraMap R (End R p) (μ₁ - μ₂) := by ext; simp [sub_smul]
+  rw [hf₁₂, IsNilpotent.map_iff (NoZeroSMulDivisors.algebraMap_injective _ _),
+    isNilpotent_iff_eq_zero, sub_eq_zero] at this
+  contradiction
+
+lemma disjoint_iSup_generalizedEigenspace [NoZeroSMulDivisors R M]
+    (f : End R M) {μ₁ μ₂ : R} (hμ : μ₁ ≠ μ₂) :
+    Disjoint (⨆ k, f.generalizedEigenspace μ₁ k) (⨆ k, f.generalizedEigenspace μ₂ k) := by
+  simp_rw [(f.generalizedEigenspace μ₁).mono.directed_le.disjoint_iSup_left,
+    (f.generalizedEigenspace μ₂).mono.directed_le.disjoint_iSup_right]
+  exact disjoint_generalizedEigenspace f hμ
+
+lemma injOn_generalizedEigenspace [NoZeroSMulDivisors R M] (f : End R M) :
+    InjOn (fun μ ↦ ⨆ k, f.generalizedEigenspace μ k)
+      {μ | ⨆ k, f.generalizedEigenspace μ k ≠ ⊥} := by
```

Comment:
totally optional:
```suggestion
    InjOn (⨆ k, f.generalizedEigenspace · k) {μ | ⨆ k, f.generalizedEigenspace μ k ≠ ⊥} := by
```

Reply:
It saves a line break so I'm all in favour!